Improve 3D Map View movement performance

dvdkon opened the pull request:
## Description

This PR improves the performance of the 3D map view, specifically removing some large stutters when moving the camera.

Currently, all work associated with camera movement (incl. finding which chunks should be visible and loading them) is done in the mouse movement event handler. This is an issue, since such events can come 1000 times a second with standard mice. These events accumulate and overrun the main thread, resulting in stutters.

With this PR, I've changed the event handler to only do the necessary event-specific computations, while leaving the rest to be done at the start of the frame.

I've also added some caching to `QgsCameraController::sampleDepthBuffer`, which sometimes computes the average depth over the whole depth buffer. This was also happening on every mouse move event and took >20ms on my machine (though release builds are likely to be faster due to inlining, I haven't checked). In this PR, I've added a cache, so this only happens once per frame.

I've also added some new `QgsEventTracing` trace points which were helpful with analysing performance and finding issues.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

dvdkon committed:
Only emit cameraChanged() on frame start

Previously the entire camera change was handled on every input event,
which led to slowdown and stutters, since we can get ~1000/s such events
with common mice. We don't care about intermediate camera states and
can only process the last state before rendering.

dvdkon committed:
Cache depth buffer average to improve performance

Previously the average could be computed on every mouse move event. This
is a problem, since the loop can be slow, especially on debug builds
where nothing gets inlined.

dvdkon committed:
Add tracing events to Qgs3DMapScene and QgsCameraController

wonder-sk reviewed (approved):
Awesome! :man_dancing:

dvdkon committed:
Fix tests after 3D camera changes

Part of the camera change handling now happens only once per frame, so
we need to wait for a new frame to be rendered so the changes are
applied.

dvdkon committed:
Fix more tests

nyalldawson commented on src/3d/qgs3dutils.cpp:
```diff
@@ -58,6 +59,23 @@ typedef Qt3DCore::QBuffer Qt3DQBuffer;
+  QEventLoop evLoop;
+  QObject::connect( frameAction, &Qt3DLogic::QFrameAction::triggered, &evLoop, &QEventLoop::quit );
+  evLoop.exec();
+  scene->removeComponent( frameAction );
```
```suggestion
  scene->removeComponent( frameAction );
  frameAction->deleteLater();
```

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -242,24 +249,28 @@ double QgsCameraController::sampleDepthBuffer( const QImage &buffer, int px, int
-  for ( int x = 0; x < buffer.width(); ++x )
+  for ( int x = 0; x < mDepthBufferImage.width(); ++x )
   {
-    for ( int y = 0; y < buffer.height(); ++y )
+    for ( int y = 0; y < mDepthBufferImage.height(); ++y )
     {
-      double d = Qgs3DUtils::decodeDepth( buffer.pixel( x, y ) );
+      double d = Qgs3DUtils::decodeDepth( mDepthBufferImage.pixel( x, y ) );
```
Ooooo we can speed this up a LOT! See https://doc.qt.io/qt-6/qimage.html#pixel "This function is expensive when used for massive pixel manipulations. Use [constBits](https://doc.qt.io/qt-6/qimage.html#constBits)() or [constScanLine](https://doc.qt.io/qt-6/qimage.html#constScanLine)() when many pixels needs to be read.". 

I'd suggest using the approach from eg QgsTerrainDownloader::tileImageToHeightMap or QgsCompoundColorWidget::averageColor instead

dvdkon replied:
That's a good idea, thanks! I've rewritten the code to use `constScanLine()`.

dvdkon committed:
Apply suggestions per review

dvdkon committed:
Speed up QgsCameraController::sampleDepthBuffer() averaging

benoitdm-oslandia commented on tests/src/3d/testqgs3dcameracontroller.cpp:
```diff
@@ -1173,5 +1172,20 @@ void TestQgs3DCameraController::testRotationCenterRotationCameraRotationCenter()
+
+    // this call is not used but ensures to synchronize the scene
+    Qgs3DUtils::captureSceneImage( engine, scene );
+  } while ( scene->cameraController()->camera()->nearPlane() < atLeast ); //#spellok
```
May be you could add a max iteration check too avoid infinite loop or too long wait.

benoitdm-oslandia commented on src/3d/qgscameracontroller.h:
```diff
@@ -350,6 +350,9 @@ class _3D_EXPORT QgsCameraController : public QObject
     bool mDepthBufferIsReady = false;
     QImage mDepthBufferImage;
+    // -1 when unset
+    // TODO: Change to std::optional<double>
```
with not using the std::optional<double>?

dvdkon replied:
We still support building with SIP 4.x, which sadly generates some code that's incompatible with C++17. Since the bindings include (some) QGIS headers, we can't have C++17-only features in them.

I made a PR to drop support for this old version of SIP (#60291), but that's currently blocked on multiple things, so no `std::optional` for now :(

dvdkon committed:
Don't loop forever in test waiting for sensible near plane

nyalldawson commented on src/3d/qgscameracontroller.h:
```diff
@@ -350,6 +350,9 @@ class _3D_EXPORT QgsCameraController : public QObject
     QImage mDepthBufferImage;
+    // -1 when unset
+    // TODO: Change to std::optional<double>
+    double mDepthBufferNonVoidAverage;
```
```suggestion
    double mDepthBufferNonVoidAverage = -1;
```

nyalldawson committed:
Update src/3d/qgscameracontroller.h
